src/Swaps/Swapper.sol

```diff
@@ -0,0 +1,134 @@
+contract Swapper is OneInchHandler, UniswapV2Handler, UniswapV3Handler, UniswapAutoRouterHandler {
+    bytes32 public constant HANDLER_ONE_INCH = bytes32("1Inch");
+    bytes32 public constant HANDLER_UNISWAP_V2 = bytes32("UniswapV2");
+    bytes32 public constant HANDLER_UNISWAP_V3 = bytes32("UniswapV3");
+    bytes32 public constant HANDLER_UNISWAP_AUTOROUTER = bytes32("UniswapAutoRouter");
+
+    uint256 internal constant REENTRANCYLOCK_UNLOCKED = 1;
+    uint256 internal constant REENTRANCYLOCK_LOCKED = 2;
+
+    uint256 private reentrancyLock;
+
+    error Swapper_UnknownMode();
+    error Swapper_UnknownHandler();
+    error Swapper_Reentrancy();
+    error Swapper_InsufficientBalance();
+    error Swapper_PastDeadline();
+
+    // In the locked state, allow contract to call itself, but block all external calls
+    modifier externalLock() {
+        bool isExternal = msg.sender != address(this);
+
+        if (isExternal) {
+            if (reentrancyLock == REENTRANCYLOCK_LOCKED) revert Swapper_Reentrancy();
+            reentrancyLock = REENTRANCYLOCK_LOCKED;
+        }
+
+        _;
+
+        if (isExternal) reentrancyLock = REENTRANCYLOCK_UNLOCKED;
+    }
+
+    constructor(address oneInchAggregator, address uniswapRouterV2, address uniswapRouterV3, address uniswapRouter02)
+        OneInchHandler(oneInchAggregator)
+        UniswapV2Handler(uniswapRouterV2)
+        UniswapV3Handler(uniswapRouterV3)
+        UniswapAutoRouterHandler(uniswapRouter02)
+    {}
+
+    /// @inheritdoc ISwapper
+    function swap(SwapParams memory params)
+        public
+        externalLock
+    {
+        if (params.mode >= MODE_MAX_VALUE) revert Swapper_UnknownMode();
+        if (params.deadline < block.timestamp) revert Swapper_PastDeadline();
+
+        if (params.handler == HANDLER_ONE_INCH) {
+            swapOneInch(params);
+        } else if (params.handler == HANDLER_UNISWAP_V2) {
+            swapUniswapV2(params);
+        } else if (params.handler == HANDLER_UNISWAP_V3) {
+            swapUniswapV3(params);
+        } else if (params.handler == HANDLER_UNISWAP_AUTOROUTER) {
+            swapAutoRouter(params);
+        } else {
+            revert Swapper_UnknownHandler();
+        }
+
+        if (params.mode == MODE_EXACT_IN) return;
+
+        // swapping to target debt is only useful for repaying
+        if (params.mode == MODE_TARGET_DEBT) {
+            // at this point amountOut holds the required repay amount
+            repayAndDeposit(params.tokenOut, params.receiver, params.amountOut, params.account);
+        }
+
+        // return unused input token after exact output swap
+        deposit(params.tokenIn, params.vaultIn, 0, params.account);
+    }
+
+    /// @inheritdoc ISwapper
+    /// @dev in case of over-swapping to repay, pass max uint amount
+    function repay(address token, address vault, uint256 repayAmount, address account) public externalLock {
```

Comment:
nit: for consistency, rename `account` to `receiver`. do the same in other functions

Reply:
I did it this way specifically to be consistent with ISwapper, where receiver is the vault for the output